src/portable/st/stm32_fsdev/dcd_stm32_fsdev.c

```diff
@@ -302,6 +309,17 @@ void dcd_set_config (uint8_t rhport, uint8_t config_num)
 void dcd_remote_wakeup(uint8_t rhport)
 {
   (void) rhport;
+  uint32_t start;
+
+  USB->CNTR |= (uint16_t)USB_CNTR_RESUME;
+  /* Wait 1 to 15 ms */
+  /* Busy loop is bad, but the osal_task_delay() isn't implemented for the "none" OSAL */
+  start = board_millis();
```

Comment:
please remove, this API is not available at all in this level. Why should it need to be delayed for too long here ?

Reply:
The remote wakeup is required to be 1 to 15 ms.

Reply:
`board_millis()` can easily be replaced with USB_FNR, each USB FS frame is 1 ms precisely. However, we still shouldn't wait here for 15 ms. That is a huge amount of time, lots of things can happen in between. We can use the little trick provided by usbd 
https://github.com/hathach/tinyusb/blob/master/src/device/usbd_pvt.h#L59

`usbd_defer_func()`, enable the bit here, then submit an function to the usbd event queue task

`usbd_defer_func( remote_wakeup_clear_bit,  USB_FNR number, false)`

`remote_wakeup_clear_bit()` will be schedule to be invoked  by tud_task(), when it is invoked, if the USB_FNR hasn't increased enough (e.g 15 frame) then we continue to queue this. The function will be called back-to-back until the we reach our desired frame. For a fast mcu, it will call like hundreds of time, however, it doesn't blocking other task.

Reply:
ah, right, there is no SOF when suspended, almost forgot that. Let's me think of another way hmmm

Reply:
RM say we can use ESOF interrupt to time this, let's use it for this purpose, enabled that interrupt when dcd_remote_wakeup() is called, then disable it when we clear the bit. We don't need `usbd_defer_func` function at all, still working without any blocking time.

![image](https://user-images.githubusercontent.com/249515/65217194-8675b780-dadd-11e9-9f7c-4171385a7c45.png)

Reply:
Another way is adding uint8_t to dcd_remote_wakeup(), and application must pass '1' to imitate the remote and '0' to finish it ( after 1-15 ms). There maybe other mcu need this API to work with. Though like other API, I will postpone the API changes until I got to work on actual hardware requiring that.